{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import math\n",
    "from datetime import datetime\n",
    "from os import path\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "Entrez.email = \"\"\n",
    "Entrez.api_key = ''\n",
    "# register api key and input here along with the registered email\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(17597, 18)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tax_id</th>\n",
       "      <th>Org_name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>CurrentID</th>\n",
       "      <th>Status</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Aliases</th>\n",
       "      <th>description</th>\n",
       "      <th>other_designations</th>\n",
       "      <th>map_location</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>genomic_nucleotide_accession.version</th>\n",
       "      <th>start_position_on_the_genomic_accession</th>\n",
       "      <th>end_position_on_the_genomic_accession</th>\n",
       "      <th>orientation</th>\n",
       "      <th>exon_count</th>\n",
       "      <th>OMIM</th>\n",
       "      <th>Unnamed: 17</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3055</td>\n",
       "      <td>Chlamydomonas reinhardtii</td>\n",
       "      <td>2717040</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>rbcL</td>\n",
       "      <td>ChreCp049</td>\n",
       "      <td>RuBisCO large subunit</td>\n",
       "      <td>RuBisCO large subunit|ribulose-1,5-bisphosphat...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_005353.1</td>\n",
       "      <td>122490.0</td>\n",
       "      <td>123917.0</td>\n",
       "      <td>minus</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4577</td>\n",
       "      <td>Zea mays</td>\n",
       "      <td>845212</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>rbcL</td>\n",
       "      <td>ZemaCp032</td>\n",
       "      <td>ribulose-1,5-bisphosphate carboxylase/oxygenas...</td>\n",
       "      <td>ribulose-1,5-bisphosphate carboxylase/oxygenas...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_001666.2</td>\n",
       "      <td>56874.0</td>\n",
       "      <td>58304.0</td>\n",
       "      <td>plus</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4097</td>\n",
       "      <td>Nicotiana tabacum</td>\n",
       "      <td>800513</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>rbcL</td>\n",
       "      <td>NitaCp031</td>\n",
       "      <td>RuBisCO large subunit</td>\n",
       "      <td>RuBisCO large subunit|ribulose-1,5-bisphosphat...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_001879.2</td>\n",
       "      <td>57600.0</td>\n",
       "      <td>59033.0</td>\n",
       "      <td>plus</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>844754</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>rbcL</td>\n",
       "      <td>ArthCp030</td>\n",
       "      <td>ribulose-1,5-bisphosphate carboxylase/oxygenas...</td>\n",
       "      <td>ribulose-1,5-bisphosphate carboxylase/oxygenas...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>54958.0</td>\n",
       "      <td>56397.0</td>\n",
       "      <td>plus</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>39946</td>\n",
       "      <td>Oryza sativa Indica Group</td>\n",
       "      <td>4126887</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>rbcL</td>\n",
       "      <td>OrsaiCp23</td>\n",
       "      <td>ribulose-1,5-bisphosphate carboxylase/oxygenas...</td>\n",
       "      <td>ribulose-1,5-bisphosphate carboxylase/oxygenas...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_008155.1</td>\n",
       "      <td>54030.0</td>\n",
       "      <td>55484.0</td>\n",
       "      <td>plus</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   tax_id                   Org_name   GeneID  CurrentID Status Symbol  \\\n",
       "0    3055  Chlamydomonas reinhardtii  2717040          0   live   rbcL   \n",
       "1    4577                   Zea mays   845212          0   live   rbcL   \n",
       "2    4097          Nicotiana tabacum   800513          0   live   rbcL   \n",
       "3    3702       Arabidopsis thaliana   844754          0   live   rbcL   \n",
       "4   39946  Oryza sativa Indica Group  4126887          0   live   rbcL   \n",
       "\n",
       "     Aliases                                        description  \\\n",
       "0  ChreCp049                              RuBisCO large subunit   \n",
       "1  ZemaCp032  ribulose-1,5-bisphosphate carboxylase/oxygenas...   \n",
       "2  NitaCp031                              RuBisCO large subunit   \n",
       "3  ArthCp030  ribulose-1,5-bisphosphate carboxylase/oxygenas...   \n",
       "4  OrsaiCp23  ribulose-1,5-bisphosphate carboxylase/oxygenas...   \n",
       "\n",
       "                                  other_designations map_location chromosome  \\\n",
       "0  RuBisCO large subunit|ribulose-1,5-bisphosphat...          NaN        NaN   \n",
       "1  ribulose-1,5-bisphosphate carboxylase/oxygenas...          NaN        NaN   \n",
       "2  RuBisCO large subunit|ribulose-1,5-bisphosphat...          NaN        NaN   \n",
       "3  ribulose-1,5-bisphosphate carboxylase/oxygenas...          NaN        NaN   \n",
       "4  ribulose-1,5-bisphosphate carboxylase/oxygenas...          NaN        NaN   \n",
       "\n",
       "  genomic_nucleotide_accession.version  \\\n",
       "0                          NC_005353.1   \n",
       "1                          NC_001666.2   \n",
       "2                          NC_001879.2   \n",
       "3                          NC_000932.1   \n",
       "4                          NC_008155.1   \n",
       "\n",
       "   start_position_on_the_genomic_accession  \\\n",
       "0                                 122490.0   \n",
       "1                                  56874.0   \n",
       "2                                  57600.0   \n",
       "3                                  54958.0   \n",
       "4                                  54030.0   \n",
       "\n",
       "   end_position_on_the_genomic_accession orientation  exon_count  OMIM  \\\n",
       "0                               123917.0       minus         0.0   NaN   \n",
       "1                                58304.0        plus         0.0   NaN   \n",
       "2                                59033.0        plus         0.0   NaN   \n",
       "3                                56397.0        plus         0.0   NaN   \n",
       "4                                55484.0        plus         0.0   NaN   \n",
       "\n",
       "   Unnamed: 17  \n",
       "0          NaN  \n",
       "1          NaN  \n",
       "2          NaN  \n",
       "3          NaN  \n",
       "4          NaN  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Inspect the datafile\n",
    "## the accession datafile can be downloaded from https://www.ncbi.nlm.nih.gov/gene/?term=rbcl\n",
    "\n",
    "acc_file = 'rbcl_acc.txt'\n",
    "seq_acc = pd.read_csv(acc_file, sep='\\t')\n",
    "print(seq_acc.shape)\n",
    "seq_acc.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Retrieve Gene seq and taxonomy from GenBank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "acc_file = 'rbcl_acc.txt'\n",
    "gene_name = 'RBCL'\n",
    "## change accession datafile name and gene name\n",
    "\n",
    "path_fastafile = 'GbRefgene/gb'+gene_name+'.fasta'\n",
    "\n",
    "\n",
    "accs = pd.read_csv(acc_file, sep='\\t')\n",
    "t0 = datetime.now()\n",
    "seqfasta = []\n",
    "ncaccs = []\n",
    "hsapiens = 0\n",
    "unknownpos = []\n",
    "\n",
    "for idx, eachrow in accs.iterrows():\n",
    "    ncacc    = eachrow['genomic_nucleotide_accession.version']\n",
    "    print(idx, 'ncacc =', ncacc)\n",
    "    if type(ncacc) == float:    continue\n",
    "    \n",
    "    startpos = eachrow['start_position_on_the_genomic_accession']\n",
    "    endpos   = eachrow['end_position_on_the_genomic_accession']\n",
    "    if math.isnan(startpos) or math.isnan(endpos):\n",
    "        unknownpos.append(ncacc)\n",
    "        continue\n",
    "    \n",
    "    \n",
    "    handle = Entrez.efetch(db=\"nucleotide\", id=ncacc, rettype=\"gb\", retmode=\"xml\")\n",
    "    gbrec  = Entrez.read(handle)\n",
    "    gbrec  = gbrec[0]\n",
    "    if 'Homo sapiens' in gbrec['GBSeq_organism']:\n",
    "        hsapiens += 1\n",
    "        continue\n",
    "        \n",
    "    handle = Entrez.efetch(db=\"nucleotide\", id=ncacc, rettype=\"fasta\", retmode=\"text\")\n",
    "    ftrec  = SeqIO.read(handle, 'fasta')\n",
    "    \n",
    "    startpos = int(startpos) - 1\n",
    "    endpos   = int(endpos)\n",
    "    ftrec.seq = ftrec.seq[startpos:endpos]\n",
    "    ftrec.description = gbrec['GBSeq_taxonomy'].replace(' ', '') + ';' + gbrec['GBSeq_organism'].replace(' ', '_')\n",
    "    \n",
    "    \n",
    "    while ncacc in ncaccs:\n",
    "        ncacc = ncacc[:-1] + str(int(ncacc[-1])+1)\n",
    "    ftrec.id = ncacc\n",
    "    ncaccs.append(ncacc)\n",
    "    seqfasta.append(ftrec)\n",
    "    \n",
    "    if (idx+1) % 1000 == 0:\n",
    "        print('Row', idx, '- Number of seq =', len(seqfasta))\n",
    "\n",
    "\n",
    "t1 = datetime.now()\n",
    "print('Total time:', t1-t0)\n",
    "print('Num of Homo sapiens =', hsapiens)\n",
    "print('Num of seq with no position =', len(unknownpos))\n",
    "SeqIO.write(seqfasta, path_fastafile, 'fasta')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Prepare local Gene local BLAST database (only need to build once)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 08/10/2021 17:41:53\n",
      "New DB name:   /home/soonjye/Documents/Snowleopard_Github/GbRefgene/gbRBCL.fasta\n",
      "New DB title:  COX3ref\n",
      "Sequence type: Nucleotide\n",
      "Keep Linkouts: T\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 13801 sequences in 0.751754 seconds.\n"
     ]
    }
   ],
   "source": [
    "## change -title\n",
    "## make sure the gene name is the same as previous\n",
    "gene_name = 'RBCL'\n",
    "\n",
    "path_fastafile = 'GbRefgene/gb'+gene_name+'.fasta'\n",
    "dbtitle = gene_name+'ref'\n",
    "!makeblastdb -in {path_fastafile} -parse_seqids -title {dbtitle} -dbtype nucl\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
